### PR TITLE
	Added support for force parameter on docker image removal

### PR DESCRIPTION
**Pull request by mjacques**
When an image is tagged in multiple repositories, image removal can only be done if the force parameter is provided.

### CODE REVIEW

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImage.groovy [hunk 1]
```diff
@@ -15,10 +15,27 @@
...
 package com.bmuschko.gradle.docker.tasks.image
 
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.Optional
+
 class DockerRemoveImage extends DockerExistingImage {
+
+	/**
+     * Forces removal.
+     */
```
Comment (cdancy) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImage.groovy]: Think we can remove this comment as the option itself is self-explanatory.

**Comment by cdancy** [comment]
@mjacques thanks for the PR! Can we have a test for this and/or add the option to an existing test?

**Comment by mjacques** [comment]
I'll add a test for docker image removal as none exists yet, and will cover up the particular case of an image tagged in multiple repositories.

**Comment by mjacques** [comment]
Is it good enough?

File: src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                force = true
+                targetImageId { buildImage.getImageId() }
+            }
+        """
+
+        when:
+        BuildResult result = build('removeImage')
+
+        then:
+        result.output.contains("Removing image with ID")
```
Comment (orzeh) [inline src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy]: Are you sure that this test actually works? In `DockerRemoveImage` logging of _"Removing image with ID"_ is done __before__ executing remove command in docker engine.

It would be better to execute another task of type `DockerListImages` and assert that target image is not in the result.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImage.groovy [hunk 3]
```diff
@@ -15,10 +15,24 @@
...
     void runRemoteCommand(dockerClient) {
         logger.quiet "Removing image with ID '${getImageId()}'."
-        dockerClient.removeImageCmd(getImageId()).exec()
+        def removeImageCmd = dockerClient.removeImageCmd(getImageId())
+        
+        if(getForce()) {
+            removeImageCmd.withForce(getForce())
+        }
+        
+        removeImageCmd.exec()
```
Comment (orzeh) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImage.groovy]: In case of `boolean` flags I think we can collapse this into one line:
````groovy
dockerClient.removeImageCmd(getImageId()).withForce(getForce()).exec()
````

File: src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy (hunk 2 above)
Comment (mjacques) [inline src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy]: I couldn't test it actually work in my dev environment because I don't have a docker server running. I was counting on continuous integration tests, but as far as I can see, functional tests don't run in there. I would appreciate some support in this area, if someone could run and adapt the tests for me.

Comment (orzeh) [inline src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy]: to spin up Docker Engine locally you can use vagrant box provided with this project:
````bash
./gradlew functionalTest \
   --tests "com.bmuschko.gradle.docker.tasks.image.DockerRemoveImageFunctionalTest*"  \
   -PbootstrapDocker=true
````

File: src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy [hunk 4]
```diff
@@ -74,13 +81,20 @@ class DockerRemoveImageFunctionalTest extends AbstractFunctionalTest {
...
+            
         """
 
         when:
-        BuildResult result = build('removeImage')
+        BuildResult result = build('removeImageAndCheckRemoval')
 
         then:
-        result.output.contains("Removing image with ID")
+        !result.output.contains("repository")
```
Comment (orzeh) [inline src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy]: It would be better not to check console log but use [custom implementation of ResponseHandler](https://github.com/bmuschko/gradle-docker-plugin/blob/master/src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerListImages.groovy#L51) which will do the check.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImage.groovy [hunk 5]
```diff
@@ -15,10 +15,21 @@
...
 package com.bmuschko.gradle.docker.tasks.image
 
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.Optional
+
 class DockerRemoveImage extends DockerExistingImage {
+
+    @Input
+    @Optional
+    Boolean force
```
Comment (cdancy) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImage.groovy]: Lets default to false as that's what docker users would expect.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImage.groovy [hunk 6]
```diff
@@ -15,10 +15,21 @@
...
+
     @Override
     void runRemoteCommand(dockerClient) {
         logger.quiet "Removing image with ID '${getImageId()}'."
-        dockerClient.removeImageCmd(getImageId()).exec()
+        def removeImageCmd = dockerClient.removeImageCmd(getImageId())
```
Comment (cdancy) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImage.groovy]: hmmm ... @orzeh if we one-line this as you noted I believe we would have to do something like this: `dockerClient.removeImageCmd(getImageId()).withForce(getForce() ? getForce().booleanValue() : false).exec()` 

Otherwise it's possible the user could set this to `null` which is not going to make `docker-java` happy. I don't think groovy will attempt to massage the `null` into `false` for java libraries. Assuming I'm right it may look cleaner to do things the previous way.

File: src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        """
+
+        when:
+        BuildResult result = build('removeImageAndCheckRemoval')
+
+        then:
+        !result.output.contains("repository")
+    }
+
+
```
Comment (cdancy) [inline src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy]: Remove added new-lines.

File: src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+                dependsOn tagImageSecondTime
+                force = true
+                targetImageId { buildImage.getImageId() }
+            }
+
+			task removeImageAndCheckRemoval(type: DockerListImages) {
+				dependsOn removeImage
+    			showAll = true
+    			filters = '{"dangling":["true"]}'
+			}
```
Comment (cdancy) [inline src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy]: Fix indentation.

File: src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        buildFile << """
+            import com.bmuschko.gradle.docker.tasks.image.Dockerfile
+            import com.bmuschko.gradle.docker.tasks.image.DockerBuildImage
+            import com.bmuschko.gradle.docker.tasks.image.DockerListImages
+            import com.bmuschko.gradle.docker.tasks.image.DockerRemoveImage
+            import com.bmuschko.gradle.docker.tasks.image.DockerTagImage
+
+            task dockerfile(type: Dockerfile) {
+                from 'ubuntu:12.04'
+            }
```
Comment (cdancy) [inline src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy]: Take a look at the other tests to see how use our "minimal file" instead of having to create another task.

File: src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+            task removeImage(type: DockerRemoveImage) {
+                dependsOn buildImage
+                targetImageId { buildImage.getImageId() }
+            }
+            
+            task removeImageAndCheckRemoval(type: DockerListImages) {
+				dependsOn removeImage
+    			showAll = true
+    			filters = '{"dangling":["true"]}'
+			}
```
Comment (cdancy) [inline src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy]: Fix indentation.

File: src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy (hunk 4 above)
Comment (cdancy) [inline src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy]: @orzeh most of our tests do this so I'm not sure if we should tell folks not to. There is an open PR/idea that I put together previously, that you commented on, that I thought would be better to implement over this. I guess my thinking is that if we are going to move towards that anyway (which I think is better) lets not tell users to use this.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImage.groovy (hunk 6 above)
Comment (orzeh) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImage.groovy]: @cdancy you are right. other solution would be to write `getForce()` as:
````groovy
Boolean getForce() {
    force ?: false
}
````

File: src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy (hunk 2 above)
Comment (mjacques) [inline src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy]: I installed vagrant and virtual box and get the following error. Any clue where the issue come from?

FAILURE: Build failed with an exception.
What went wrong:
VirtualBox is not functional. Please check!

Comment (cdancy) [inline src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy]: @mjacques do you have the full stacktrace?

Comment (orzeh) [inline src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy]: What VirtualBox version? If 5.1 it might not work, see [vagrant issue](https://github.com/mitchellh/vagrant/issues/7411)

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImage.groovy (hunk 6 above)
Comment (cdancy) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImage.groovy]: @orzeh @mjacques lets stick with what we/mjacques did previously. It's more verbose but inline with what we do on other tasks and is very clear to the user what the intent is.

File: src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy (hunk 2 above)
Comment (mjacques) [inline src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy]: @cdancy @orzeh 
I tried with virtualbox 5.1, 5.0, 4.3 and still same issue. I have also added virtualbox folder to my path, it didn't help.
Vagrant on his own does work well.

Here is the full stacktrace:

> org.gradle.api.GradleException: VirtualBox is not functional. Please check!
>         at com.bmuschko.gradle.vagrant.VagrantBasePlugin$_validateVagrantInstallation_closure3.doCall(VagrantBasePlugin.groovy:76)
>         at org.gradle.listener.ClosureBackedMethodInvocationDispatch.dispatch(ClosureBackedMethodInvocationDispatch.java:40)
>         at org.gradle.listener.ClosureBackedMethodInvocationDispatch.dispatch(ClosureBackedMethodInvocationDispatch.java:25)
>         at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:44)
>         at org.gradle.internal.event.BroadcastDispatch.dispatch(BroadcastDispatch.java:79)
>         at org.gradle.internal.event.BroadcastDispatch.dispatch(BroadcastDispatch.java:30)
>         at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
>         at com.sun.proxy.$Proxy52.graphPopulated(Unknown Source)
>         at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter.execute(DefaultTaskGraphExecuter.java:108)
>         at org.gradle.execution.SelectedTaskExecutionAction.execute(SelectedTaskExecutionAction.java:37)
>         at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:37)
>         at org.gradle.execution.DefaultBuildExecuter.access$000(DefaultBuildExecuter.java:23)
>         at org.gradle.execution.DefaultBuildExecuter$1.proceed(DefaultBuildExecuter.java:43)
>         at org.gradle.execution.DryRunBuildExecutionAction.execute(DryRunBuildExecutionAction.java:32)
>         at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:37)
>         at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:30)
>         at org.gradle.initialization.DefaultGradleLauncher$4.run(DefaultGradleLauncher.java:153)
>         at org.gradle.internal.Factories$1.create(Factories.java:22)
>         at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:91)
>         at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:53)
>         at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:150)
>         at org.gradle.initialization.DefaultGradleLauncher.access$200(DefaultGradleLauncher.java:32)
>         at org.gradle.initialization.DefaultGradleLauncher$1.create(DefaultGradleLauncher.java:98)
>         at org.gradle.initialization.DefaultGradleLauncher$1.create(DefaultGradleLauncher.java:92)
>         at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:91)
>         at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:63)
>         at org.gradle.initialization.DefaultGradleLauncher.doBuild(DefaultGradleLauncher.java:92)
>         at org.gradle.initialization.DefaultGradleLauncher.run(DefaultGradleLauncher.java:83)
>         at org.gradle.launcher.exec.InProcessBuildActionExecuter$DefaultBuildController.run(InProcessBuildActionExecuter.java:99)
>         at org.gradle.tooling.internal.provider.ExecuteBuildActionRunner.run(ExecuteBuildActionRunner.java:28)
>         at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
>         at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:48)
>         at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:30)
>         at org.gradle.launcher.exec.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:81)
>         at org.gradle.launcher.exec.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:46)
>         at org.gradle.launcher.exec.DaemonUsageSuggestingBuildActionExecuter.execute(DaemonUsageSuggestingBuildActionExecuter.java:51)
>         at org.gradle.launcher.exec.DaemonUsageSuggestingBuildActionExecuter.execute(DaemonUsageSuggestingBuildActionExecuter.java:28)
>         at org.gradle.launcher.cli.RunBuildAction.run(RunBuildAction.java:43)
>         at org.gradle.internal.Actions$RunnableActionAdapter.execute(Actions.java:173)
>         at org.gradle.launcher.cli.CommandLineActionFactory$ParseAndBuildAction.execute(CommandLineActionFactory.java:239)
>         at org.gradle.launcher.cli.CommandLineActionFactory$ParseAndBuildAction.execute(CommandLineActionFactory.java:212)
>         at org.gradle.launcher.cli.JavaRuntimeValidationAction.execute(JavaRuntimeValidationAction.java:35)
>         at org.gradle.launcher.cli.JavaRuntimeValidationAction.execute(JavaRuntimeValidationAction.java:24)
>         at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:33)
>         at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:22)
>         at org.gradle.launcher.cli.CommandLineActionFactory$WithLogging.execute(CommandLineActionFactory.java:205)
>         at org.gradle.launcher.cli.CommandLineActionFactory$WithLogging.execute(CommandLineActionFactory.java:169)
>         at org.gradle.launcher.Main.doAction(Main.java:33)
>         at org.gradle.launcher.bootstrap.EntryPoint.run(EntryPoint.java:45)
>         at org.gradle.launcher.bootstrap.ProcessBootstrap.runNoExit(ProcessBootstrap.java:55)
>         at org.gradle.launcher.bootstrap.ProcessBootstrap.run(ProcessBootstrap.java:36)
>         at org.gradle.launcher.GradleMain.main(GradleMain.java:23)
>         at org.gradle.wrapper.BootstrapMainStarter.start(BootstrapMainStarter.java:30)
>         at org.gradle.wrapper.WrapperExecutor.execute(WrapperExecutor.java:127)
>         at org.gradle.wrapper.GradleWrapperMain.main(GradleWrapperMain.java:56)
>

**Comment by cdancy** [comment]
Merged this PR as I have a need for its functionality ;) I'll address the comments before I release this weekend. Thanks @mjacques !!!

**Comment by mjacques** [comment]
@cdancy You're welcome! Sorry I couldn't address the comments in an effective manner.
